pkg/services/datasources/service/datasource_service.go

```diff
@@ -296,10 +303,18 @@ func (s *Service) DecryptedValues(ctx context.Context, ds *models.DataSource) (m
 		err = json.Unmarshal([]byte(secret), &decryptedValues)
 	}
 
-	if (!exist || err != nil) && len(ds.SecureJsonData) > 0 {
-		decryptedValues, err = s.MigrateSecrets(ctx, ds)
-		if err != nil {
-			return nil, err
+	if len(ds.SecureJsonData) > 0 {
+		if !exist || err != nil {
+			decryptedValues, err = s.MigrateSecrets(ctx, ds)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if s.features.IsEnabled("disableSecretsCompatibility") {
+			err := s.DeleteDataSourceSecrets(ctx, &models.DeleteDataSourceSecretsCommand{UID: ds.Uid, OrgID: ds.OrgId})
```

Comment:
not necessary but for completeness

```suggestion
			err := s.DeleteDataSourceSecrets(ctx, &models.DeleteDataSourceSecretsCommand{UID: ds.Uid, OrgID: ds.OrgId, ID: ds.Id})
```

Reply:
Currently looking on how to improve this.

Reply:
@marefr Wouldn't the decryption cache lock be enough to prevent race conditions related to the DecryptedValues since the first request would lock the cache and after that the secret would always be retrieved from cache?

Reply:
Which decryption cache lock are you referring to, a new one or the existing one in the secrets store?

Reply:
I have converted the goroutine into a background service on 35ed14e8

But now I’m wondering, should I leave this on the datasources service? Or should I move it to the secrets kvstore since new migrations might be implemented in the future for other things to be integrated with unified secrets?

Reply:
Sorry, thought I answered this. Moving it probably makes more sense.